### PR TITLE
Thread Safety Fix

### PR DESCRIPTION
**Pull request by radar**
This is an alternative take on #51, using concurrent-ruby instead of the thread_safe gem.

### CODE REVIEW

**Comment by thedarkone** [comment]
@radar, `ThreadSafe::Cache` (with the rest of `thread_safe`) has been been merged into `concurrent-ruby`. It is called `Concurrent::Map` now.

Re the PR: you should be using `Concurrent::Map` instead of `Concurrent::Hash` (formerly `ThreadSafe::Hash`). Admittedly this is confusing, but as rule of thumb one should always pick `Conc::Map` over `Conc::Hash` unless insertion order is important.

What about all other places that I swapped `Hash` for `TS::Cache` in the original PR?

File: lib/i18n.rb [hunk 1]
```diff
@@ -335,7 +337,7 @@ def normalize_key(key, separator)
     end
 
     def normalized_key_cache
-      @normalized_key_cache ||= Hash.new { |h,k| h[k] = {} }
+      @normalized_key_cache ||= Concurrent::Hash.new { |h,k| h[k] = Concurrent::Hash.new }
```
Comment (radar) [inline lib/i18n.rb]: @romuloceccon Could you please submit a PR to my branch here to fix that?

Comment (romuloceccon) [inline lib/i18n.rb]: Sure, and sorry for the delay. I'll do it during the weekend. Meanwhile, here's a demonstration of why the code above is not thread safe: https://gist.github.com/romuloceccon/f44a30cb43f8081279a484193d386f55.

File: i18n.gemspec [hunk 2]
```diff
@@ -19,4 +19,6 @@ Gem::Specification.new do |s|
   s.rubyforge_project = '[none]'
   s.required_rubygems_version = '>= 1.3.5'
   s.required_ruby_version = '>= 1.9.3'
+
+  s.add_dependency 'concurrent-ruby', '1.0.2'
```
Comment (kares) [inline i18n.gemspec]: could we, eventually, not have the version locked down so specifically but instead e.g. `~> 1.0`

**Comment by kares** [comment]
ah I see, this is simply stale for almost a year, what's the hold up anything this needs help with?

File: i18n.gemspec (hunk 2 above)
Comment (radar) [reply i18n.gemspec]: Relaxed in 5480dd33298dd11b24667f47862f8a063a9fad6f.

**Comment by radar** [comment]
@kares I assume you're interested in this because you have a problem that would be fixed by either this PR or one very close to it. So based on that assumption, I'm going to ask you if you'd be available to test this PR to see if it would solve your issues. Please do try it.

Thanks!

**Comment by radar** [comment]
I'm targeting this and #382 for a 0.9 release, probably in the next week or two, assuming that both get tested / approved by then. So it would be really appreciated if you could test this, @kares.

**Comment by kares** [comment]
Thanks Ryan, I expected some but than for now it turned out to be something else. 
Felt scary that this has been identified for so long and not handled right, I think we can come up with a threaded test-case to emulate a potential concurrent issue.

**Comment by radar** [comment]
With #384 done, I think this can be merged now.

**Pull request merged**